Create Rent Model

cquiro opened the pull request:
## Summary
- Create rent model.
- Add has_many :rents line to book and user models.
- Create rents factory.
- Create users factory.
- Create empty rent_spec.rb file.

## Trello Card
https://trello.com/c/Do93K2LL/21-crear-modelo-de-rent

cquiro committed:
Create rent model

- Create rent model.
- Add has_many :rents line to book and user models.

cquiro committed:
Create spec files for rent.

- Create rents factory.
- Create users factory.
- Create empty rent_spec.rb file.

cquiro committed:
Fix rubocop offenses

- Add dependent: :destroy to has_many line in book and user.
- Replace double quotes with single quotes in rents and users factories.

lmasello commented on db/migrate/20180924124835_create_rents.rb:
```diff
@@ -0,0 +1,12 @@
+    create_table :rents do |t|
+      t.references :user, foreign_key: true
+      t.references :book, foreign_key: true
+      t.date :start_date
```
Since `start_date` and `end_date` are required fields, don't forget to add the null: false statement.
Also don't forget to validate the presence in the model.

lmasello commented on spec/factories/rents.rb:
```diff
@@ -0,0 +1,8 @@
+  factory :rent do
+    user
+    book
+    start_date { '2018-09-24' }
```
Use Faker::Date https://github.com/stympy/faker/blob/master/doc/date.md

lmasello commented on spec/factories/rents.rb:
```diff
@@ -0,0 +1,8 @@
+    user
+    book
+    start_date { '2018-09-24' }
+    end_date { '2018-09-24' }
```
Try to set the end_date relative to the random start_date selected by Faker. This could be, for example 5 days after the start_date.

lmasello commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,8 @@
+    first_name { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    email { Faker::Internet.email }
+    password { 'password' }
```
Don't you like Faker::Internet.password(8)?

lmasello commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Rent, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
If we add presence validators, we should test them.

cquiro committed:
Add specs for rent

- Write specs for rent
- Write validator in rent to make sure end_date is never earlier than
start date.

facucompe commented on app/models/user.rb:
```diff
@@ -1,4 +1,6 @@
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
+
+  has_many :rents, dependent: :destroy
```
Be careful with `dependent: :destroy`. In this case, you may want to keep the rents although the user no longer exists.
Maybe `:nullify`?

facucompe commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  it 'is not valid without a user' do
+    rent = build(:rent, user: nil)
+
+    expect(rent).not_to be_valid
```
Same as the previous PR. `it { is_expected.to validate_presence_of(:user) }`

facucompe commented on spec/models/rent_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  it 'is not valid when end_date < start_date' do
+    rent = build(:rent, start_date: Time.current, end_date: 2.days.ago)
+    rent.valid?
```
Unnecessary line

cquiro replied:
@facucompe I tried using create instead of build, but the test fails before getting to the expectations. Using .valid? was the only way I found to trigger the validations while also being able to test my custom error message (build does not trigger validations). Could you please point me in the right direction?

facucompe replied:
I think you should be able to do this:
```
subject { build(:rent, start_date: Time.current, end_date: 2.days.ago) }
it { is_expected.not_to be_valid }
```

cquiro committed:
Merge branch test_book_model into create_rent_model.

cquiro committed:
Use Faker gem in factories

- Use Faker::Date in rents factory.

- Use Faker::Internet.password(8) in users factory.

cquiro committed:
Use one-liners to test validations in rent_spec.rb

cquiro committed:
Validate for presence of rent start and end dates

- Specify 'null: false' in the create rents migration.

- Reset db.

- Remove code to validate presence of user and book in the
rent model.

cquiro committed:
Test custom validation for success

In rent_spec.rb test if custom validation works, not if the error
message returned is as expected.
